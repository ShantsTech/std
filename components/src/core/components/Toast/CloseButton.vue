<!--
/*
 * This file is part of Shants Tech LLC
 *
 * Copyright (C) 2020 Shants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="std-toast-close-container">
    <div :class="classes" role="button">
      Ã—
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';

export default defineComponent({
  name: 'std-toast-close-button',

  mixins: [toastMixin],

  computed: {
    classes(): object {
      return {
        'std-toast-close': true,
        [`std-toast-close--${this.type}`]: true,
      };
    },
  },
});
</script>

<style src="./close-button.scss" lang="scss" scoped></style>
