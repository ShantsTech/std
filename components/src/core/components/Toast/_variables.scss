// This file is part of Shants Tech LLC
//
// Copyright (C) 2020 Shants Tech LLC
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

$std-toast-icon-color-default: $std-interface-gray-darken-1-color !default;
$std-toast-icon-color: $std-white-color !default;
$std-toast-font-color-default: $std-interface-gray-darken-1-color !default;
$std-toast-font-color: $std-white-color !default;
$std-toast-border-radius: 3.2rem !default;
$std-toast-padding: 0.6rem !default;
$std-toast-box-shadow: 0px 25px 35px 0 rgba(0, 0, 0, 0.18) !default;
$std-toast-opacity: 0.8 !default;
$std-toast-content-padding-horizontal: 0.8rem !default;

$std-toast-bg-color-default: $std-white-color !default;
$std-toast-bg-color-success: $std-feedback-success-color !default;
$std-toast-bg-color-warn: $std-feedback-warn-color !default;
$std-toast-bg-color-error: $std-feedback-danger-color !default;
$std-toast-bg-color-info: $std-feedback-info-color !default;

$std-toast-icon-wrap-size: 3.2rem !default;
$std-toast-icon-wrap-padding: 1rem !default;
$std-toast-icon-font-size: 1.5rem !default;

$std-toast-close-size: 1rem !default;
$std-toast-close-padding: 0.2rem !default;
$std-toast-close-border-radius: 1rem !default;

$std-toast-container-space-between: 0.4rem !default;
