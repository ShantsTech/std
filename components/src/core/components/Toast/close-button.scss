// This file is part of Shants Tech LLC
//
// Copyright (C) 2020 Shants Tech LLC
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

@import '../../../styles';
@import 'variables';

.std-toast-close {
  border: 0;
  outline: 0;
  padding: $std-toast-close-padding;
  height: $std-toast-close-size;
  width: $std-toast-close-size;
  border-radius: $std-toast-close-border-radius;
  cursor: pointer;
  background-color: rgba($std-toast-icon-color, 0.1);
  @include std-flex-center;

  &--default {
    background-color: rgba($std-toast-icon-color-default, 0.1);
  }
}

.std-toast-close-container {
  @include std-flex-center;
}
