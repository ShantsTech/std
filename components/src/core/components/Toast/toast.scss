// This file is part of Shants Tech LLC
//
// Copyright (C) 2020 Shants Tech LLC
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

@import '../../../styles';
@import 'variables';

.std-toast {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  color: $std-toast-icon-color;
  box-shadow: $std-toast-box-shadow;
  padding: $std-toast-padding;
  border-radius: $std-toast-border-radius;
  opacity: $std-toast-opacity;
  cursor: pointer;

  &:hover {
    opacity: 1;
  }

  &--default {
    color: $std-toast-icon-color-default;
    background-color: $std-toast-bg-color-default;
  }

  &--success {
    background-color: $std-toast-bg-color-success;
  }

  &--warn {
    background-color: $std-toast-bg-color-warn;
  }

  &--error {
    background-color: $std-toast-bg-color-error;
  }

  &--info {
    background-color: $std-toast-bg-color-info;
  }

  &--persistent {
    opacity: 1;
    cursor: unset;
  }
}

.std-toast-start {
  display: flex;
  flex-direction: row;
}

.std-toast-content {
  display: flex;
  flex-direction: column;
  padding-left: $std-toast-content-padding-horizontal;
  padding-right: $std-toast-content-padding-horizontal;
  color: $std-toast-font-color;

  &--default {
    color: $std-toast-font-color-default;
  }
}

.std-toast-content-text {
  color: inherit;
}
