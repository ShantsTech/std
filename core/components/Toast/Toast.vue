<!--
/*
 * This file is part of Stants Tech LLC
 *
 * Copyright (C) 2020 Stants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div v-if="show" :class="classes" aria-live="assertive" @click="onClickToast">
    <div class="std-toast-start">
      <slot name="icon">
        <std-toast-icon :type="type" />
      </slot>
      <slot>
        <div :class="contentClasses">
          <std-text type="toast-title" class="std-toast-content-text">
            {{ title }}
          </std-text>
          <std-text type="toast-message" class="std-toast-content-text">
            {{ message }}
          </std-text>
        </div>
      </slot>
    </div>
    <slot name="close">
      <std-toast-close-button :type="type" @click="onClickClose" />
    </slot>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';
import Text from '@sthrm/std/core/components/Text/Text.vue';
import ToastCloseButton from '@sthrm/std/core/components/Toast/CloseButton.vue';
import ToastIcon from '@sthrm/std/core/components/Toast/Icon.vue';

export default defineComponent({
  name: 'std-toast',

  mixins: [toastMixin],

  components: {
    'std-text': Text,
    'std-toast-close-button': ToastCloseButton,
    'std-toast-icon': ToastIcon,
  },

  emits: ['update:show'],

  props: {
    show: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      required: true,
    },
    message: {
      type: String,
    },
    persistent: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    classes(): object {
      return {
        'std-toast': true,
        [`std-toast--${this.type}`]: true,
        ['std-toast--persistent']: this.persistent,
      };
    },
    contentClasses(): object {
      return {
        'std-toast-content': true,
        [`std-toast-content--${this.type}`]: true,
      };
    },
  },

  methods: {
    onClickToast() {
      if (!this.persistent) {
        this.$emit('update:show', false);
      }
    },
    onClickClose(e: Event) {
      e.stopPropagation();
      this.$emit('update:show', false);
    },
  },
});
</script>

<style src="./toast.scss" lang="scss" scoped></style>
