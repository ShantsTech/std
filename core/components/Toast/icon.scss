// This file is part of Stants Tech LLC
//
// Copyright (C) 2020 Stants Tech LLC
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

@import '../../../styles';
@import 'variables';

.std-toast-icon {
  color: inherit;
  font-size: $std-toast-icon-font-size;
}

.std-toast-icon-container {
  @include std-flex-center;
}

.std-toast-icon-wrap {
  background-color: rgba($std-toast-icon-color, 0.1);
  padding: $std-toast-icon-wrap-padding;
  border-radius: 100%;
  box-sizing: border-box;
  height: $std-toast-icon-wrap-size;
  width: $std-toast-icon-wrap-size;
  @include std-flex-center;

  &--default {
    background-color: rgba($std-toast-icon-color-default, 0.1);
  }
}
