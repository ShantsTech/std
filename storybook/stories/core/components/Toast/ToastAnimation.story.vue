<!--
/*
 * This file is part of Shants Tech LLC
 *
 * Copyright (C) 2020 Shants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->
-->

<template>
  <div>
    <button @click="show = !show">
      Toggle
    </button>

    <transition name="std-toast-fade">
      <std-toast v-model:show="show" title="Success" />
    </transition>
  </div>
</template>

<script>
import Toast from '@sthrm/std/core/components/Toast/Toast.vue';

export default {
  components: {
    'std-toast': Toast,
  },
  data() {
    return {
      show: true,
    };
  },
};
</script>

<style lang="scss" scoped>
.std-toast-fade {
  // https://v3.vuejs.org/guide/transitions-enterleave.html#transitioning-single-elements-components
  &-enter-active {
    transition: opacity 0.5s ease-out;
  }
  &-enter-from {
    opacity: 0;
  }

  &-leave-active {
    transition: opacity 1s cubic-bezier(1, 0.5, 0.8, 1);
  }
  &-leave-to {
    opacity: 0 !important;
  }
}
</style>
