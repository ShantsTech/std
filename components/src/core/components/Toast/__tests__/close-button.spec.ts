/*
 * This file is part of Shants Tech LLC
 *
 * Copyright (C) 2020 Shants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount} from '@vue/test-utils';
import Toast from '@sthrm/std/core/components/Toast/CloseButton.vue';

describe('Toast > CloseButton.vue', () => {
  it('should renders STD ToastCloseButton', () => {
    const wrapper = mount(Toast, {
      props: {},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders STD ToastCloseButton success', () => {
    const wrapper = mount(Toast, {
      props: {type: 'success'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders STD ToastCloseButton warn', () => {
    const wrapper = mount(Toast, {
      props: {type: 'warn'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders STD ToastCloseButton error', () => {
    const wrapper = mount(Toast, {
      props: {type: 'error'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders STD ToastCloseButton info', () => {
    const wrapper = mount(Toast, {
      props: {type: 'info'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
