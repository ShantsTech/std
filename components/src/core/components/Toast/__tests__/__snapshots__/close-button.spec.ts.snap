// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast > CloseButton.vue should renders STD ToastCloseButton 1`] = `
<div class="std-toast-close-container">
  <div class="std-toast-close std-toast-close--default" role="button"> × </div>
</div>
`;

exports[`Toast > CloseButton.vue should renders STD ToastCloseButton error 1`] = `
<div class="std-toast-close-container">
  <div class="std-toast-close std-toast-close--error" role="button"> × </div>
</div>
`;

exports[`Toast > CloseButton.vue should renders STD ToastCloseButton info 1`] = `
<div class="std-toast-close-container">
  <div class="std-toast-close std-toast-close--info" role="button"> × </div>
</div>
`;

exports[`Toast > CloseButton.vue should renders STD ToastCloseButton success 1`] = `
<div class="std-toast-close-container">
  <div class="std-toast-close std-toast-close--success" role="button"> × </div>
</div>
`;

exports[`Toast > CloseButton.vue should renders STD ToastCloseButton warn 1`] = `
<div class="std-toast-close-container">
  <div class="std-toast-close std-toast-close--warn" role="button"> × </div>
</div>
`;
